_posts/2021-03-19-forecasting-time-series-with-torch_4/forecasting_4.Rmd

```diff
@@ -0,0 +1,570 @@
+
+Here, we show two implementation options, one additive, and one multiplicative.
+
+#### Additive variant
+
+In additive attention, encoder outputs and decoder state are commonly either added or concatenated (we choose to do the latter, below). The resulting tensor is run through a linear layer, and a softmax is applied for normalization.
+
+```{r}
+attention_module_additive <- nn_module(
+  
+  initialize = function(hidden_dim, attention_size) {
+    
+    self$attention <- nn_linear(2 * hidden_dim, attention_size)
+    
+  },
+  
+  forward = function(state, encoder_outputs) {
+    
+    # function argument shapes
+    # encoder_outputs: (bs, timesteps, hidden_dim)
+    # state: (1, bs, hidden_dim)
+    
+    # multiplex state to allow for concatenation (dimensions 1 and 2 must agree)
+    seq_len <- dim(encoder_outputs)[2]
+    # resulting shape: (bs, timesteps, hidden_dim)
+    state_rep <- state$permute(c(2, 1, 3))$repeat_interleave(seq_len, 2)
+    
+    # concatenate along feature dimension
+    concat <- torch_cat(list(state_rep, encoder_outputs), dim = 3)
+    
+    # run through linear layer with tanh
+    # resulting shape: (bs, timesteps, attention_size)
+    scores <- self$attention(concat) %>% 
+      torch_tanh()
+    
+    # sum over attention dimension and normalize
+    # resulting shape: (bs, timesteps) 
+    attention_weights <- scores %>%
+      torch_sum(dim = 3) %>%
+      nnf_softmax(dim = 2)
+    
+    # a normalized score for every source token
+    attention_weights
+  }
+)
+
+```
+
+#### Multiplicative variant
+
+In multiplicative attention, scores are obtained by computing dot products between decoder state and all of the encoder outputs. Here too, a softmax is then used for normalization.
+
+```{r}
+attention_module_multiplicative <- nn_module(
+  
+  initialize = function() {
+    
+    NULL
+    
+  },
+  
+  forward = function(state, encoder_outputs) {
+    
+    # function argument shapes
+    # encoder_outputs: (bs, timesteps, hidden_dim)
+    # state: (1, bs, hidden_dim)
+
+    # allow for matrix multiplication with encoder_outputs
+    state <- state$permute(c(2, 3, 1))
+ 
+    # prepare for scaling by number of features
+    d <- torch_tensor(dim(encoder_outputs)[3], dtype = torch_float())
+       
+    # scaled dot products between state and outputs
+    # resulting shape: (bs, timesteps, 1)
+    scores <- torch_bmm(encoder_outputs, state) %>%
+      torch_div(torch_sqrt(d))
+    
+    # normalize
+    # resulting shape: (bs, timesteps) 
+    attention_weights <- scores$squeeze(3) %>%
+      nnf_softmax(dim = 2)
+    
+    # a normalized score for every source token
+    attention_weights
+  }
+)
+```
+
+### Decoder
+
+Once attention weights have been computed, their actual application is handled by the decoder. Concretely, the aptly named method `weighted_encoder_outputs()` computes a product of weights and encoder outputs, making sure that each output will have appropriate impact.
```

Comment:
```suggestion
Once attention weights have been computed, their actual application is handled by the decoder. Concretely, the aptly named method, `weighted_encoder_outputs()`, computes a product of weights and encoder outputs, making sure that each output will have appropriate impact.
```

Reply:
>Concretely, the aptly named method, `weighted_encoder_outputs()`, computes

I think that with the commas, I need to add something there, perhaps
> Concretely, the aptly named method in question, `weighted_encoder_outputs()`, computes

Reply:
no, wait ... why not just

> Concretely, the method in question, weighted_encoder_outputs(), computes